tests/integration_tests/test_public_api.py

```diff
@@ -0,0 +1,144 @@
+
+
+@contextmanager
+def launch_instance(
+    cloud_instance: BaseCloud, **kwargs
+) -> Generator[BaseInstance, None, None]:
+    instance = cloud_instance.launch(**kwargs)
+    try:
+        yield instance
+    finally:
+        instance.delete()
+
+
+def assert_example_output(instance: BaseInstance):
+    example_output = instance.execute("cat /var/tmp/example.txt").stdout
+    assert example_output == "hello"
+
+
+def exercise_push_pull(instance: BaseInstance):
+    with TemporaryDirectory() as tmpdir:
+        push_path = Path(tmpdir).joinpath("pushed")
+        push_path.write_text("pushed", encoding="utf-8")
+        instance.push_file(str(push_path), "/var/tmp/pushed")
+        assert (
+            "pushed" == instance.execute("cat /var/tmp/pushed").stdout.strip()
+        )
+
+        instance.execute("echo 'pulled' > /var/tmp/pulled")
+        pull_path = Path(tmpdir).joinpath("pulled")
+        instance.pull_file("/var/tmp/pulled", str(pull_path))
+        assert pull_path.read_text(encoding="utf-8").strip() == "pulled"
+
+
+def exercise_instance(instance: BaseInstance):
+    with suppress(NotImplementedError):
+        assert instance.name is not None
+
+    with suppress(NotImplementedError):
+        assert ipaddress.ip_address(instance.ip)
+
+    with suppress(NotImplementedError):
+        ip_address = instance.add_network_interface()
+        try:
+            assert ipaddress.ip_address(ip_address)
+        finally:
+            instance.remove_network_interface(ip_address)
+
+    exercise_push_pull(instance)
+
+    assert_example_output(instance)
+    instance.execute("sync")
+    boot_id = instance.get_boot_id()
+    instance.restart()
+    assert boot_id != instance.get_boot_id()
+
+    assert_example_output(instance)
+    instance.shutdown()
+
+    instance.start()
+    assert_example_output(instance)
+
+
+@pytest.mark.parametrize(
+    "cloud",
+    [
+        pytest.param(pycloudlib.Azure, id="azure"),
+        pytest.param(pycloudlib.EC2, id="ec2"),
+        pytest.param(pycloudlib.GCE, id="gce"),
+        pytest.param(pycloudlib.IBM, id="ibm"),
+        pytest.param(pycloudlib.LXDContainer, id="lxd_container"),
+        pytest.param(pycloudlib.LXDVirtualMachine, id="lxd_vm"),
+        pytest.param(pycloudlib.OCI, id="oci"),
+        # For openstack we first need a reliable way of obtaining the
+        # image id
+        # pytest.param(pycloudlib.Openstack, id="openstack"),
+    ],
+    indirect=True,
+)
+def test_public_api(cloud: BaseCloud):
+    """Shallow test of (most) public functions in the base API."""
+    latest_lts = LTS_RELEASES[-1]
+    try:
+        image_id = cloud.released_image(release=latest_lts)
+    except NotImplementedError:
+        image_id = cloud.daily_image(release=latest_lts)
+    with suppress(NotImplementedError):
+        # Not sure there's a great way to test this other than not raising
+        cloud.image_serial(image_id)
+
+    with launch_instance(
```

Comment:
Since lxd doesn't observe the cloud.tag in instance naming, can we standardize this instance naming before launch?
```diff
diff --git a/tests/integration_tests/test_public_api.py b/tests/integration_tests/test_public_api.py
index 262b16b..a4f272b 100644
--- a/tests/integration_tests/test_public_api.py
+++ b/tests/integration_tests/test_public_api.py
@@ -129,8 +129,9 @@ def test_public_api(cloud: BaseCloud):
         # Not sure there's a great way to test this other than not raising
         cloud.image_serial(image_id)
 
+    name = cloud.tag + f"-{latest_lts}"
     with launch_instance(
-        cloud, image_id=image_id, user_data=cloud_config, wait=False
+        cloud, image_id=image_id, user_data=cloud_config, wait=False, name=name
     ) as instance:
         instance.wait()
         exercise_instance(instance)
```

Reply:
Heh, I actually already have a branch ready to fallback to tag if `name` isn't defined. Was gonna wait for this one to merge before putting it up, but I can put it up now-ish. You ok leaving this as-is  if we add that functionality?

Reply:
#295

Reply:
Yes let's leave 295 separate from this PR thanks.